common/ring/builder.go

```diff
@@ -837,6 +865,37 @@ func (b *RingBuilder) adjustReplica2Part2DevSize(toAssign map[uint][]uint) {
 func (b *RingBuilder) gatherPartsFromFailedDevices(assignParts map[uint][]uint) int {
 	// First we gather partitions from removed devices.  Since removed devices usually indicate device failures, we have no choice but to reassing these partitions.  However, we mark them as moves so later choices will skip other replicas of the same partition if possible.
 	// TODO: Implement this to support removing devices
+	if len(b.removedDevs) > 0 {
+		devsWithParts := make([]*RingBuilderDevice, 0)
+		for i, dev := range b.removedDevs {
+			if dev.Parts > 0 {
+				devsWithParts = append(devsWithParts, b.removedDevs[i])
+			}
+		}
+		if len(devsWithParts) > 0 {
+			for replica, part2Dev := range b.replica2Part2Dev {
+				for part, devId := range part2Dev {
+					for _, dev := range b.removedDevs {
```

Comment:
Should this be range devsWithParts? Also, since that's all devsWithParts is really for, maybe make it a []uint from the start?